subsys/bluetooth/host/hci_core.c

```diff
@@ -317,7 +317,7 @@ int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
 
 	BT_DBG("opcode 0x%04x status 0x%02x", opcode, cmd(buf)->status);
 
-	if (cmd(buf)->status) {
+	if (err || cmd(buf)->status) {
```

Comment:
There's a assert for `err == 0` just a few lines above so adding a separate check doesn't make sense. Either there should be the assert or what you propose. Originally HCI_CMD_TIMEOUT didn't exist, rather this waited with K_FOREVER. The timeout was added to catch bugs with the implementation and/or controller. The assert was originally chosen since If we fail to get a command complete or status we loose sync with the command flow control. I'm not 100% sure it's safe to change it to a runtime error check - this needs to be properly verified, however in general this feels like a workaround. If you controller has the ability to be "disconnected" there should be some out-of-band signal between the driver and the host to notify of this, rather than through arbitrary timeouts.

Reply:
Well, I have no easy way to check that since the controller has only a UART connection to the MCU, though I could probably check CTS manually before running bt_init. But I agree this patch looks like a workaround. However, reporting success in case of timeout (when CONFIG_ASSERT disabled) also means that HCI 'sync loose' is not detected at all.

Reply:
@loicpoulain one clean way to do something like this is with `CONFIG_BT_WAIT_NOP`. If you can make your controller emit a command_complete(NOP) when it's done initializing then with the help of the mentioned Kconfig option you can make the Zephyr host stack wait for that event before commencing any other communication with the controller. E.g. the Zephyr controller is able to be configured to do this (same Kconfig option in that case). Some older CSR controllers also had a pskey that enabled this kind of functionality.

Reply:
> @loicpoulain one clean way to do something like this is with `CONFIG_BT_WAIT_NOP`. If you can make your controller emit a command_complete(NOP) when it's done initializing then with the help of the mentioned Kconfig option you can make the Zephyr host stack wait for that event before commencing any other communication with the controller. E.g. the Zephyr controller is able to be configured to do this (same Kconfig option in that case). Some older CSR controllers also had a pskey that enabled this kind of functionality.

@jhedberg It looks like a good suggestion indeed. I was not aware of this NOP CC event, and it seems to be totally aligned with the spec:

> To indicate to the Host that the Controller is ready to receive HCI command packets, the Controller generates a Command Complete event with the Command_Opcode set to 0x0000 and the Num_HCI_Command_Packets event parameter set to 1 or more...

Thanks!

Loic